### PR TITLE
[Relay][Frontend] Caffe2 Support

### PR DESCRIPTION
**Pull request by makihiro**
@hlu1 @tqchen @zhreshold @kazum Please help to review.

### CODE REVIEW

**Comment by yzhliu** [comment]
Please satisfy the CI
```
from model_zoo import c2_squeezenet, c2_resnet50, c2_vgg19

  File "/workspace/tests/python/frontend/caffe2/model_zoo/__init__.py", line 19, in <module>

    locals()['c2_' + model] = importlib.import_module('caffe2.python.models.' + model)

  File "/usr/lib/python3.6/importlib/__init__.py", line 126, in import_module

    return _bootstrap._gcd_import(name[level:], package, level)

ModuleNotFoundError: No module named 'caffe2.python.models.squeezenet'
```

File: tests/python/frontend/caffe2/test_forward.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    # here we use the first input blob to the first op to get the input name
+    input_names = model.predict_net.op[0].input[0]
+    shape_dict = {input_names: input_data.shape}
+    dtype_dict = {input_names: input_data.dtype}
+    func, params = relay.frontend.from_caffe2(model.init_net, model.predict_net, shape_dict, dtype_dict)
+    with relay.build_config(opt_level=3):
+        graph, lib, params = relay.build(func, target, params=params)
+
+    ctx = tvm.cpu(0)
```
Comment (yzhliu) [inline tests/python/frontend/caffe2/test_forward.py]: can we use `tvm.relay.testing.config.ctx_list` to test it on all supported targets?

File: tests/python/frontend/caffe2/test_forward.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+
+def verify_vgg19():
+    verify_caffe2_forward_impl(c2_vgg19, (1, 3, 224, 224), (1, 1000))
+
+
+if __name__ == '__main__':
+    verify_squeezenet1_1()
+    verify_resnet50()
+    verify_vgg19()
```
Comment (kazum) [inline tests/python/frontend/caffe2/test_forward.py]: Test methods should start with "test_" so that nosetests can execute them.

**Comment by yzhliu** [comment]
@makihiro could you act on the comments?

**Comment by makihiro** [comment]
> @makihiro could you act on the comments?

Thank you for the comments, and sorry for the late reply.
Module import error occurred due to no caffe2 models installed on CI environment.
The test code tries to download models and install into caffe2 path, however there is permission error for this operation.
The best solution is caffe2 models are pre-installed on CI environment because total model size is several hundreds of megabytes.
To fix this, I'm going to modify to download models into temporary directory (eg. /tmp) and load for now.

File: tests/python/frontend/caffe2/test_forward.py (hunk 1 above)
Comment (makihiro) [reply tests/python/frontend/caffe2/test_forward.py]: ctx comes from ctx_list and I'm going to remove ctx = tvm.cpu(0)

File: tests/python/frontend/caffe2/test_forward.py (hunk 2 above)
Comment (makihiro) [reply tests/python/frontend/caffe2/test_forward.py]: Thank you for the review. I fixed them.

File: tests/python/frontend/caffe2/test_forward.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+
+def test_verify_vgg19():
+    verify_caffe2_forward_impl(c2_vgg19, (1, 3, 224, 224), (1, 1000))
+
+
+if __name__ == '__main__':
+    test_verify_squeezenet1_1()
+    test_verify_resnet50()
+    test_verify_vgg19()
```
Comment (kazum) [inline tests/python/frontend/caffe2/test_forward.py]: Replace 'test_verify' with 'test_forward' to be consistent with other frontend tests.

File: tests/python/frontend/caffe2/model_zoo/__init__.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    return model_dir
+
+
+# skip download if model exist
+for model in models:
+    try:
+        locals()['c2_' + model] = importlib.import_module('caffe2.python.models.' + model)
+    except ImportError:
+        try:
+            os.system("python3 -m caffe2.python.models.download -i -f " + model)
```
[the remark this answers is not in the document]
Comment (makihiro) [reply tests/python/frontend/caffe2/model_zoo/__init__.py]: Thank you for the review. I fixed code to use ModelDownloader

File: tests/python/frontend/caffe2/test_forward.py (hunk 3 above)
Comment (makihiro) [reply tests/python/frontend/caffe2/test_forward.py]: I renamed functions in a manner similar to other frontend tests.

**Pull request merged**